specification/under_development/Base-Profile.md

```diff
@@ -0,0 +1,160 @@
+
+## Program Structure
+
+A Base Profile compliant program is defined in the form of a single LLVM bitcode file that consists of the following:
+- definition of the `Qubit` and `Result` types
+- [entry point definition](#entry-point-definition)
+- declarations of [qis functions](#quantum-instruction-set)
+- declarations of functions used for [output recording](#output-recording)
+- [attributes](#attributes)
+
+The human readable LLVM IR for the bitcode can be obtained using standard 
+[LLVM tools](https://llvm.org/docs/CommandGuide/llvm-dis.html). For the purpose of clarity, this specification contains examples of the human readable IR emitted by [LLVM 13](https://releases.llvm.org/13.0.1/docs/LangRef.html). While the bitcode reprentation is portable and usually backward compatible, there may be visual differences in the human readable format depending on the LLVM version. This differences are irrelevant when using standard tools to load, manipulate, and/or execute bitcode.
+
+The code below illustrates how a simple program looks like within a Base Profile representation:
+
+```llvm
+; type definitions
+
+%Result = type opaque
+%Qubit = type opaque
+
+; entry point definition
+
+define void @EntryPointName() #0 {
+entry:
+
+  ; calls to qis functions  
+  tail call void @__quantum__qis__h__body(%Qubit* null)
+  tail call void @__quantum__qis__cnot__body(%Qubit* null, %Qubit* inttoptr (i64 1 to %Qubit*))
+  tail call void @__quantum__qis__mz__body(%Qubit* null, %Result* null)
+  tail call void @__quantum__qis__mz__body(%Qubit* inttoptr (i64 1 to %Qubit*), %Result* inttoptr (i64 1 to %Result*))
+
+  ; calls to record the program output
+  tail call void @__quantum__rt__tuple_start_record_output()
+  tail call void @__quantum__rt__result_record_output(%Result* null)
+  tail call void @__quantum__rt__result_record_output(%Result* inttoptr (i64 1 to %Result*))
+  tail call void @__quantum__rt__tuple_end_record_output()
+
+  ret void
+}
+
+; declarations of qis functions
+
+declare void @__quantum__qis__h__body(%Qubit*)
+
+declare void @__quantum__qis__cnot__body(%Qubit*, %Qubit*)
+
+declare void @__quantum__qis__mz__body(%Qubit*, %Result*)
+
+; declarations of functions used for output recording
+
+declare void @__quantum__rt__tuple_start_record_output()
+
+declare void @__quantum__rt__result_record_output(%Result*)
+
+declare void @__quantum__rt__tuple_end_record_output()
+
+; attributes
+
+attributes #0 = { "EntryPoint" "requiredQubits"="2" "requiredResults"="2" }
+```
+
+TODO: a profile identifier and version number within the IR would be good to have (may be needed for correct usage of qubit and result pointers)   
+-> look into custom target triples
```

Comment:
Could we consider a new identifying mechanism that is _similar_ to a target triple, but that doesn't overload the existing target, e.g. `target triple = "x86_64-apple-macosx10.7.0"`? 

There is a use case where an executable binary compiled from bitcode containing QIR could be executed on a classical architecture (and quantum instructions are sent e.g. over a network to a QPU), and to lose the original target triple from a front-end would likely cause compilation problems in a later stage. 

Not sure if Attribute Groups are suitable for this, but something that doesn't remove any other target information would be excellent. Apologies if this was not the expressed intent, but I figured I'd share a thought!

Reply:
Yes, that's a good point. I would still like to better understand the opportunities that having LLVM recognize targeting information might give us, but unless there is a good reason, I agree to leave the LLVM target triple and find another representation for the QIR specific targeting information.

Reply:
In the most recent write up, we now have an attribute identifying the profile, and module flags identifying the spec version. It does not seem to make sense to identify the instruction set at this time - this is implicit by the declarations.